### PR TITLE
Naturalist Lens

### PR DESCRIPTION
**Pull request by pspjuth**
This adds a Naturalist Lens to help localise where National Parks can be built.
It extends the Naturalist from MoreLenses by highlighting potential tiles that
could become a Park except for something simple such as removing an
improvement or change ownership.

### CODE REVIEW

**Comment by pspjuth** [comment]
@bolbass, I don' t understand your 2).
Regarding 1), I prefer to know if bordering tiles are worthy. I'm not above stealing a tile for a good park.
However, buying a neutral tile is quite different from conquering one, so having another hue for those owned by someone else might be prudent.

**Comment by bolbass** [comment]
@pspjuth, regarding 2): in your method any tile with >=2 appeal will be highlighted as a "ok" or "fixable" tile even if it's surrounded with 6 districts and can never be used for National Park. That's why first we need to check if all 4 tiles are "ok" or "fixable" and then highlight all of them in that case.
regarding 1): it's a bit strange to see "fixable" tiles near other civ's capitals or other cities. I think that conquering these cities is much harder "fix" than adding neutral territory to your civ.

**Comment by pspjuth** [comment]
@bolbass, I see what you mean and I agree that would be ideal.

However, I think it now has a good balance between complexity and usefulness. The current implementation is dead simple since it analyses single tiles, and very useful since it is trivial to disregard the impossibilities and see the diamonds.

Extending the analysis to look at neighboring tiles would create a new category of "fixable", where appeal can be increased by removing mines, planting forests etc. The list of possibilities there is rather large and
getting it fully right would be tricky.  But I have some ideas running through my mind, I'll see what I can
do.

**Comment by pspjuth** [comment]
@bolbass, With this addition the clutter should be less while still providing good information.

File: Integrations/ML/UI/minimappanel.lua [hunk 1]
```diff
@@ -1451,45 +1451,97 @@ function SetNaturalistLens()
...
-      good = true;
+      data.Level = 3;
     elseif pPlot:GetAppeal() >= 2 then
-      good = true;
+      -- Charming
+      data.Level = 3;
+    elseif pPlot:GetAppeal() >= 1 then
+      -- It is tricky to analyse if a lower appeal is fixable so we
+      -- view it as semifixable without further analysis.
+      data.Level = 1;
```
Comment (bolbass) [inline Integrations/ML/UI/minimappanel.lua]: @pspjuth, you don't use these plots later. Anyway I think that it's better not to count plots with appeal < 2 at all. It's enough to have 2 states only: good or improve, like it was before.

File: Integrations/ML/UI/minimappanel.lua [hunk 2]
```diff
@@ -1451,45 +1451,97 @@ function SetNaturalistLens()
...
+    local p3:table = Map.GetPlot(data.X, data.Y + 2);
+
+    -- All three must exist
+    if p1 ~= nil and p2 ~= nil and p3 ~= nil then
+      local i1 = p1:GetIndex();
+      local i2 = p2:GetIndex();
+      local i3 = p3:GetIndex();
+      -- All three must have stored data
+      if tiles[i1] ~= nil and tiles[i2] ~= nil and tiles[i3] ~= nil then
+        -- At least one must be owned
```
Comment (bolbass) [inline Integrations/ML/UI/minimappanel.lua]: I think we need to show diamonds that are fully on neutral territory also.

File: Integrations/ML/UI/minimappanel.lua [hunk 3]
```diff
@@ -1451,45 +1451,97 @@ function SetNaturalistLens()
...
-    if plotHasImprovement(pPlot) then
-      improve = true;
-    end
-    -- Ownership can change
-    if pPlot:GetOwner() ~= Game.GetLocalPlayer() then
-      improve = true;
+      data.Level = 0;
+    end
+
+    -- Only keep relevant tiles
```
Comment (bolbass) [inline Integrations/ML/UI/minimappanel.lua]: you need also to exclude plots that are in rawParkPlots table or you will color the same plots twice.

File: Integrations/ML/UI/minimappanel.lua [hunk 4]
```diff
@@ -1451,45 +1451,97 @@ function SetNaturalistLens()
...
   local fixableHexes:table = {};
   local okHexes:table = {};
+  local tiles:table = {};
 
+  -- Collect individual tile data
   for i = 0, (mapWidth * mapHeight) - 1, 1 do
     local pPlot:table = Map.GetPlotByIndex(i);
-    local improve:boolean = false;
-    local good:boolean = false;
+    local data =  {
```
Comment (bolbass) [inline Integrations/ML/UI/minimappanel.lua]: I wouldn't count plots that aren't on neutral or player's territory at all

File: Integrations/ML/UI/minimappanel.lua [hunk 5]
```diff
@@ -1451,45 +1451,97 @@ function SetNaturalistLens()
   local mapWidth, mapHeight = Map.GetGridSize();
   local fixableHexes:table = {};
   local okHexes:table = {};
+  local tiles:table = {};
 
+  -- Collect individual tile data
   for i = 0, (mapWidth * mapHeight) - 1, 1 do
     local pPlot:table = Map.GetPlotByIndex(i);
```
Comment (bolbass) [inline Integrations/ML/UI/minimappanel.lua]: I wouldn't count plots that aren't on neutral or player's territory at all

**Comment by bolbass** [comment]
And by the way More Lenses taken to CQUI from the More Lenses mod so it's better to commit changes there also

**Comment by chaorace** [comment]
@pspjuth Do you plan to make any additional changes in response to the review?

**Comment by pspjuth** [comment]
@chaorace I definitiely plan to look more at this, I've just been totally swamped by work. Vacation is coming up so that will help :-)

**Comment by astog** [comment]
@pspjuth Haha, I also finished my Naturalist Lens version 😄. I wanted to do it a lot earlier but other things got in the way and I had to take a break from modding. I'm curious about the types of tiles you created. What is the difference _ok_ and _raw_ park plots. Also what do you mean by fixable? Is it that an improvement can be removed to make place for a National Park?

**Comment by bolbass** [comment]
@astog _raw_ park plots in @pspjuth method are the plots that you count in your method. _Ok_ park plots are the plots that are ready to become raw park plots after some "fixes" (adding neutral tiles to your territory or removing improvements for example). _Fixable_ park plots are the plots that should be "fixed" to become _ok_ park plots. If a diamond consists of 4 _ok_ hexes they become _raw_ park plots.
Here is an example:
1. pspjuth (when he will make additional changes in response to the review):
more_lenses_1
2. astog:
more_lenses_2

And by the way your colors are a bit dark for me.

**Comment by astog** [comment]
That is a lot more useful than my version. As long as @pspjuth permits, I want to use his code this lens.

**Comment by pspjuth** [comment]
Just a quick update: I will be able to take care of this mid next week.

@astog, feel free to use my code in MoreLenses if you want. 
I have some more mixed lens stuff at https://github.com/pspjuth/cqui/commits/mystuff
if you want to have a look and use any of it.

**Comment by astog** [comment]
Update, I integrated this lens in my latest [commit](https://github.com/astog/MoreLenses/commit/84d499bf84259e86e80df02b40ee5914c4b80523). @pspjuth thank you again for the original code.

**Comment by chaorace** [comment]
@pspjuth I'm closing this PR now since the changes have since been absorbed into the full More Lenses mod. Rest assured, I'll remember to credit you in the usual places when the new ML version gets integrated into CQUI

**Pull request closed without merging**